# Load important Libraries and Packages

In [1]:
#import librarbies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Machine learning required packages

In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

# Data Discribtion:
# House Price prediction with linear Algorithem:
# Data consist of to predict house price based upon given idependent variable/input Feature.
# find out which model help us to provide better result/prediction.

# Read and Explore the data

In [3]:
# Load the data
boost_df = pd.read_csv('Boston.csv')
boost_df

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio   black  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

#Data Observation:
    a.on the given data we have 13 input feature/columns and 1 target column
    b.on the given data 13 independent variable and 1 dependent variable 

In [7]:
# checking null #of rows and columns
boost_df.shape
#found 506 rows with 14 columns along with target columns

(506, 14)

In [11]:
#check for the null value
boost_df.isnull().sum()
# we dont have any null value in the given data

crim       0
zn         0
indus      0
chas       0
nox        0
rm         0
age        0
dis        0
rad        0
tax        0
ptratio    0
black      0
lstat      0
medv       0
dtype: int64

In [14]:
#Check duplciates
boost_df.duplicated().sum()
#not found any duplicates

0

In [15]:
#check for duplicate row number
boost_df[boost_df.duplicated]

Empty DataFrame
Columns: [crim, zn, indus, chas, nox, rm, age, dis, rad, tax, ptratio, black, lstat, medv]
Index: []

In [17]:
#check info
boost_df.info()
#provides number of column header with data type

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  black    506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


In [27]:
#check only data type
print(boost_df.dtypes)

crim       float64
zn         float64
indus      float64
chas         int64
nox        float64
rm         float64
age        float64
dis        float64
rad          int64
tax          int64
ptratio    float64
black      float64
lstat      float64
medv       float64
dtype: object


In [37]:
# convert specific column data type
boost_df = boost_df.astype({"lstat": float})
boost_df
print(boost_df.dtypes)

crim       float64
zn         float64
indus      float64
chas         int64
nox        float64
rm         float64
age        float64
dis        float64
rad          int64
tax          int64
ptratio    float64
black      float64
lstat      float64
medv       float64
dtype: object


In [38]:
#check if we have any NA values
boost_df.dropna()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio   black  lstat  medv  
0       15.3  396.90    4.0  24.0  
1       17.8  396.90    9.0  21.6  
2       17.8  392.83    4.0  34.7  
3       18.7  394.63    2.0  33.4  
4       18.7  396.90    5.0  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99    9.0  22.4  
502     21.0  396.90    9.0  20.6  
503     21.0  396.90    5.0  23.9  
504     21.0  393.45    6.0  22.0  
505     21.0  396.90    7.0  11.9  

[506 rows x 14 columns]

# 3. Machine Learning Process

In [46]:
#Step1:
#store all the input features/independent variable into X_train expect target column = 'medv'
X = boost_df.drop(columns = 'medv')
#store target variable/dependent variable into y_train which is 'medv'
y = boost_df['medv']
X

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio   black  lstat  
0       15.3  396.90    4.0  
1       17.8  396.90    9.0  
2       17.8  392.83    4.0  
3       18.7  394.63    2.0  
4       18.7  396.90    5.0  
..       ...     ...    ...  
501     21.0  391.99    9.0  
502     21.0  396.90    9.0  
503     21.0  396.90    5.0  
504     21.0  393.45    6.0  
505     21.0  396.90    7.0  

[506 rows x 13 columns]

In [58]:
#Split the data into training and testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

# Apply Linear Regression on the Data

In [59]:
#Define linear regression model
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

In [62]:
#check coefficient: from m1 to m13 columns:
#Coefficients help you identify which features are most important in your model. 
#Features with larger coefficients have a greater impact on the predictions.
lin_reg.coef_

array([-8.92064600e-02,  3.72228368e-02, -1.65395271e-02,  3.37971768e+00,
       -1.70063181e+01,  3.97684333e+00, -1.37987971e-03, -1.36412119e+00,
        2.60332528e-01, -1.01996992e-02, -1.05021029e+00,  1.06571503e-02,
       -4.91631995e-01])

In [63]:
#Check "C" intercept #C value / constant
lin_reg.intercept_

35.30209251689115

In [65]:
#predict y_pred with X_test to check in lin_ regression
y_pred = lin_reg.predict(X_test)
y_pred

array([45.33558795, 19.66204947, 19.93559847, 30.49711575, 22.06510883,
       25.18202156, 35.99842827, 32.47275585, 20.37807646, 36.44744739,
       24.15359652, 22.20865924, 30.96831137, 12.28574682, 15.99364338,
        9.08989398, 26.5754011 , 25.29565667, 22.43976442, 30.69875461,
       21.61517669, 21.23524893, 27.27433488, 11.03836075, 18.20230548,
       15.7953065 , 31.26779432, 17.87162555, 23.65343098, 27.75194807,
       29.46610817, 15.94218661, 26.8765078 , 38.23116503, 29.1592938 ,
       24.56867574, 35.44746961, 15.57044784, 28.23559277, 30.4988579 ,
       13.85474888, 35.43123101, 19.82834784, 25.65216318, 18.21794536,
       21.14999082, 30.8304069 , 19.59965648, 18.52392819, 25.11689942,
       18.80461572, 10.15761606, 20.34327591, 19.02071135, 22.15249504,
       25.58442006,  6.3864205 , 17.77544122, 11.61080874, 23.77817392,
       23.31721945,  9.20388134, 13.3598714 , 35.29843524, 28.74370811,
       30.32523632, 25.18675642, 24.38990779, 12.85996677, 13.28

In [66]:
#y_test actual price
y_test

283    50.0
28     18.4
122    20.5
253    42.8
45     19.3
       ... 
363    16.8
263    31.0
44     21.2
34     13.5
301    22.0
Name: medv, Length: 102, dtype: float64

# Evaluation Metrics for Regression models:

In [ ]:
1.MSE(MEAN SQUARED ERROR:)
2.RMSE(ROOT MEAN SQUARED ERROR:)
3.R-SQUARED(r2_score)
4.ADJUSTED R-SQUARED(Adjusted r2_Score)

In [69]:
#Load mean square error from sklearn
from sklearn.metrics import mean_squared_error

In [72]:
mse = mean_squared_error(y_test, y_pred)
mse

34.248421353723074

In [74]:
#ROOT MEAN SQUARED
rmse = mse ** 0.5
rmse

5.85221508095209

In [75]:
#Compare y_pred with actual value with r2_score
r2_score(y_test, y_pred)

0.5983144025200919

# Insights from Linear regression model

#Observation:
   a> r2_scre = When compared with r2 from "y_test" with "y_pred" and the output for the model prediction is "%59.83" overall which is not good % for the prediction.
   b> mse = Since r2_score and model prediction performs is not great hence mse is "34.24".
   c> rmse = Around 5.85 will be subtracted from y_pred for each iteration.

# Final conclusion from Linear model

## Linear Regression prediction from the given data is different from actual value vs predicted value by an average of 5.85 will subtratced from each input.